README.md
# Robot-Tondeuse
Robot Tondeuse Arduino

Source : https://repalmakershop.com/

![IMG_20220604_145247_resized_20220608_114639966](https://user-images.githubusercontent.com/100352663/173809259-5330e8c3-3a5f-40d9-a4be-061af2bc07e7.jpg)
![IMG_20220604_212029_resized_20220608_114638618](https://user-images.githubusercontent.com/100352663/173809281-a1c3d5af-2237-4564-84bc-a24bfb49de02.jpg)


History.txt

2022-01-09 Robot_V9.72_NoDelay_Coms_v1.zip - communication without delay()
2022-01-11 Robot_V9.72_NoDelay_Coms_v2.zip - corrected serial port on TFT
2022-01-23 Nano_2021-01-23.ino             - averaging of values
2022-01-29 Robot_V9.72_NoDelay_Coms_v3.zip - NODELAY_BACKWARD, WEBSERVER, WDT (watchdogTimer), Wheel Blocked TFT msg+Fake value test, compass lib timeout
2022-02-15 Robot_V9.72_NoDelay_Coms_v4.zip - 
  NodeMCU - MCUBLYNK_NOT_USED - posible run webserver without Blynk account, 
          - WEBSERVER - more web pages added, possible to control mower, setting not fully working yet
  Mega    - WEBSERVER - more web pages added, possible to control mower, setting not fully working yet
  TFT	  - bug with communication fixed, communication between Mega and TFT was working only on Serial3 on TFT
2022-02-26 Robot_V9.72_NoDelay_Coms_v4_1.zip -
  NodeMCU - file WebPages.ino instead of WebPages.h - corrected
  NodeMCU+MEGA - added some configuration over webserver
2022-04-02 Robot_V9.72_NoDelay_Coms_v4_2.zip - only bug fixes
  MEGA - Sonar.ino - disabled auto Sonar Off if errors - due to added timeout in pulseIn()
       - Manouvers.ino - disable WDT in Manouver_Go_To_Charging_Station() because of a lot of delay() used
  TFT  - Menu_Tests.ino - fixed Wheel test Aerator/Mower
2022-04-05 Robot_V9.72_NoDelay_Coms_v4_3.zip - only bug fixes   
  Mega - some serial output correction
       - less compass readings as loop CT is much faster than origin code
  library DFRobot_QMC5883 - added timeout message to serial output when reading does not respond (!!! HMC(QMC) Compass library timeout !!!)
2022-04-18 Robot_V9.72_NoDelay_Coms_v4_4.zip - 
  Nano - fixed calculation of RawDrillAmp and Raining value
  TFT  - fixed black screen when Rain detected
  Mega - added Mow time to Alarms - time change only over code in Activate_Alarms() (Mow_Time(x)) yet
  TFT+Mega - added timeout to confirmation of communication - sometimes Please wait... message freeze on TFT
2022-04-28 Robot_V9.72_NoDelay_Coms_v5.zip - 
  Mega+TFT - sending + setting date - preparation for calendar in alarm
  Mega - StopBtn - fixed problem with relay spikes which activate StopBtn attachinterrupt without its pressing	
  Mega - disabled serial messages regarding GPS when GPS not enabled
  Mega - WheelAmp and Tilt sensor working in backward too (only if run backward due to sonar yet)
2022-05-01 Robot_V9.72_NoDelay_Coms_v5_1.zip -
  Mega - Specials_Find_Wire_Track - disable WDT in Specials_Find_Wire_Track() because of a lot of delay() used
  Mega - added delay timer to runbackward functions
  Mega - added Fake_wire to UpdateWireSensor() and Check_Wire_In_Out()
  Mega - increase Wheel_Blocked_Count_Max when run backward in Check_Wheel_Amps()
2022-05-20 Robot_V9.72_NoDelay_Coms_v6.zip -
  SerialCom library v1.1 - working with Hardware and Software serial, no need any modification inside (=> removed #define USE_SW_SERIAL)
  Mega - Wheel_Status_Value for NodeMCU serial output 
  NodeMCU - telnet - serial output over WiFi with same data as on Mega serial output - no need USB cable for watching Mega values
                   - library TelnetSpy and Serial. renamed to SerialDebug. 
  Mega+NodeMCU - increased buffer (40->80) for SerialCom as more values are transfered
  
2022-05-30 - Robot_V9.73_NoDelay_Coms_v7.zip -
  - based on Robot_V9.731
  - corrected setting and printing of Date and Time
2022-06-05 - Robot_V9.73_NoDelay_Coms_v7_1.zip 
  - Mega - fixed reading of date when getting initial values
  - TFT - fixed wrong screen when click on Quick Go->Just Go

library/telnetspy/TelnetSpy.cpp
/*
 * TELNET SERVER FOR ESP8266 / ESP32
 * Cloning the serial port via Telnet.
 *
 * Written by Wolfgang Mattis (arduino@yasheena.de).
 * Version 1.1 / September 7, 2018. 
 * MIT license, all text above must be included in any redistribution.   
 */

#ifdef ESP8266
extern "C" {
	#include "user_interface.h"
}
#endif

#include "../telnetspy/TelnetSpy.h"

#ifndef min
#define min(a,b) ((a)<(b)?(a):(b))
#endif
#ifndef max
#define max(a,b) ((a)>(b)?(a):(b))
#endif

static TelnetSpy* actualObject = NULL;


static void TelnetSpy_putc(char c) {
	if (actualObject) {
  		actualObject->write(c);
	}
}

static void TelnetSpy_ignore_putc(char c) {;
}

TelnetSpy::TelnetSpy() {
	port = TELNETSPY_PORT;
	telnetServer = NULL;
	started = false;
	listening = false;
	firstMainLoop = true;
	usedSer = &Serial;
	storeOffline = true;
	connected = false;
	callbackConnect = NULL;
	callbackDisconnect = NULL;
	welcomeMsg = strdup(TELNETSPY_WELCOME_MSG);
	rejectMsg = strdup(TELNETSPY_REJECT_MSG);
	minBlockSize = TELNETSPY_MIN_BLOCK_SIZE;
	collectingTime = TELNETSPY_COLLECTING_TIME;
	maxBlockSize = TELNETSPY_MAX_BLOCK_SIZE;
	pingTime = TELNETSPY_PING_TIME;
	pingRef = 0xFFFFFFFF;
	waitRef = 0xFFFFFFFF; 
	telnetBuf = NULL;
	bufLen = 0;
	uint16_t size = TELNETSPY_BUFFER_LEN;
	while (!setBufferSize(size)) {
		size = size >> 1;
		if (size < minBlockSize) {
			setBufferSize(minBlockSize);
			break;
		}
	}
	debugOutput = TELNETSPY_CAPTURE_OS_PRINT;
	if (debugOutput) {
		setDebugOutput(true);
	}
}

TelnetSpy::~TelnetSpy() {
	end();
}

void TelnetSpy::setPort(uint16_t portToUse) {
	port = portToUse;
	if (listening) {
		if (client.connected()) {
			client.flush();
			client.stop();
		}
		if (connected && (callbackDisconnect != NULL)) {
			callbackDisconnect();
		}
		connected = false;
		telnetServer->close();
		delete telnetServer;
		telnetServer = new WiFiServer(port);
		if (started) {
			telnetServer->begin();
			telnetServer->setNoDelay(bufLen > 0);
		}
	}
}

void TelnetSpy::setWelcomeMsg(char* msg) {
	if (welcomeMsg) {
		free(welcomeMsg);
	}
	welcomeMsg = strdup(msg);
}

void TelnetSpy::setRejectMsg(char* msg) {
	if (rejectMsg) {
		free(rejectMsg);
	}
	rejectMsg = strdup(msg);
}

void TelnetSpy::setMinBlockSize(uint16_t minSize) {
	minBlockSize = min(max((uint16_t) 1, minSize), maxBlockSize);
}
    
void TelnetSpy::setCollectingTime(uint16_t colTime) {
	collectingTime = colTime;
}

void TelnetSpy::setMaxBlockSize(uint16_t maxSize) {
	maxBlockSize = max(maxSize, minBlockSize);
}

bool TelnetSpy::setBufferSize(uint16_t newSize) {
	if (telnetBuf && (bufLen == newSize)) {
		return true;
	}
	if (newSize == 0) {
		bufLen = 0;
		if (telnetBuf) {
			free(telnetBuf);
			telnetBuf = NULL;
		}
		if (telnetServer) {
			telnetServer->setNoDelay(false);
		}
		return true;
	}
	newSize = max(newSize, minBlockSize);
	uint16_t oldBufLen = bufLen;
	bufLen = newSize;
	uint16_t tmp;
	if (!telnetBuf || (bufUsed == 0)) {
		bufRdIdx = 0;
		bufWrIdx = 0;
		bufUsed = 0;
	} else {
		if (bufLen < oldBufLen) {
			if (bufRdIdx < bufWrIdx) {
				if (bufWrIdx > bufLen) {
					tmp = min(bufLen, (uint16_t) (bufWrIdx - max(bufLen, bufRdIdx)));
					memcpy(telnetBuf, &telnetBuf[bufWrIdx - tmp], tmp);
					bufWrIdx = tmp;
					if (bufWrIdx > bufRdIdx) {
						bufRdIdx = bufWrIdx;
					} else {
						if (bufRdIdx > bufLen) {
							bufRdIdx = 0;
						}
					}
					if (bufRdIdx == bufWrIdx) {
						bufUsed = bufLen;
					} else {
						bufUsed = bufWrIdx - bufRdIdx;
					}
				}
			} else {
				if (bufWrIdx > bufLen) {
					memcpy(telnetBuf, &telnetBuf[bufWrIdx - bufLen], bufLen);
					bufRdIdx = 0;
					bufWrIdx = 0;
					bufUsed = bufLen;
				} else {
					tmp = min(bufLen - bufWrIdx, oldBufLen - bufRdIdx);
					memcpy(&telnetBuf[bufLen - tmp], &telnetBuf[oldBufLen - tmp], tmp);
					bufRdIdx = bufLen - tmp;
					bufUsed = bufWrIdx + tmp;
				}
			}
		}
	}
	char* temp = (char*) realloc(telnetBuf, bufLen);
	if (!temp) {
		return false;
	}
	telnetBuf = temp;
	if (telnetBuf && (bufLen > oldBufLen) && (bufRdIdx > bufWrIdx)) {
		tmp = bufLen - (oldBufLen - bufRdIdx);
		memcpy(&telnetBuf[tmp], &telnetBuf[bufRdIdx], oldBufLen - bufRdIdx);
		bufRdIdx = tmp;
	}
	if (telnetServer) {
		telnetServer->setNoDelay(true);
	}
	return true;
}

uint16_t TelnetSpy::getBufferSize() {
	if (!telnetBuf) {
		return 0;
	}
	return bufLen;
}

void TelnetSpy::setStoreOffline(bool store) {
	storeOffline = store;
}

bool TelnetSpy::getStoreOffline() {
	return storeOffline;
}

void TelnetSpy::setPingTime(uint16_t pngTime) {
	pingTime = pngTime;
	if (pingTime == 0) {
		pingRef = 0xFFFFFFFF;
	} else {
		pingRef = (millis() & 0x7FFFFFF) + pingTime;
	}
}

void TelnetSpy::setSerial(HardwareSerial* usedSerial) {
	usedSer = usedSerial;
}

size_t TelnetSpy::write (uint8_t data) {
	if (telnetBuf) {
		if (storeOffline || client.connected()) {
			if (bufUsed == bufLen) {
				if (client.connected()) {
					sendBlock();
				}
				if (bufUsed == bufLen) {
					char c;
					while (bufUsed > 0) {
						c = pullTelnetBuf();
						if (c == '\n') {
							break;
						}
					}
					if (peekTelnetBuf() == '\r') {
						pullTelnetBuf();
					}
				}
			}
			addTelnetBuf(data);
		}
	} else {
		if (client.connected()) {
			client.write(data);
		}
	}
	if (usedSer) {
		return usedSer->write(data);
	}
	return 1;
}
    
int TelnetSpy::available (void) {
	if (usedSer) {
		int avail = usedSer->available();
		if (avail > 0) {
			return avail;
		}
	}
	if (client.connected()) {
		return telnetAvailable();
	}
	return 0;
}

int TelnetSpy::read (void) {
	int val;
	if (usedSer) {
		val = usedSer->read();
		if (val != -1) {
			return val;
		}
	}
	if (client.connected()) {
		if (telnetAvailable()) {
			val = client.read();
		}
	}
	return val;
}
    
int TelnetSpy::peek (void) {
	int val;
	if (usedSer) {
		val = usedSer->peek();
		if (val != -1) {
			return val;
		}
	}
	if (client.connected()) {
		if (telnetAvailable()) {
			val = client.peek();
		}
	}
	return val;
}
    
void TelnetSpy::flush (void) {
	if (usedSer) {
		usedSer->flush();
	}
}

#ifdef ESP8266

void TelnetSpy::begin(unsigned long baud, SerialConfig config, SerialMode mode, uint8_t tx_pin) {
	if (usedSer) {
		usedSer->begin(baud, config, mode, tx_pin);
	}
	started = true;
}

#else // ESP32

void TelnetSpy::begin(unsigned long baud, uint32_t config, int8_t rxPin, int8_t txPin, bool invert) {
	if (usedSer) {
		usedSer->begin(baud, config, rxPin, txPin, invert);
	}
	started = true;
}

#endif

void TelnetSpy::end() {
	if (debugOutput) {
		setDebugOutput(false);
	}
	if (usedSer) {
		usedSer->end();
	}
	if (client.connected()) {
		client.flush();
		client.stop();
	}
	if (connected && (callbackDisconnect != NULL)) {
		callbackDisconnect();
	}	
	connected = false;
	telnetServer->close();
	delete telnetServer;
	telnetServer = NULL;
	listening = false;
	started = false;
}

#ifdef ESP8266

void TelnetSpy::swap(uint8_t tx_pin) {
	if (usedSer) {
		usedSer->swap(tx_pin);
	}
}

void TelnetSpy::set_tx(uint8_t tx_pin) {
	if (usedSer) {
		usedSer->set_tx(tx_pin);
	}
}

void TelnetSpy::pins(uint8_t tx, uint8_t rx) {
	if (usedSer) {
		usedSer->pins(tx, rx);
	}
}

bool TelnetSpy::isTxEnabled(void) {
	if (usedSer) {
		return usedSer->isTxEnabled();
	}
	return true;
}

bool TelnetSpy::isRxEnabled(void) {
	if (usedSer) {
		return usedSer->isRxEnabled();
	}
	return true;
}

#endif

int TelnetSpy::availableForWrite(void) {
	if (usedSer) {
		return min(usedSer->availableForWrite(), bufLen - bufUsed);
	}
	return bufLen - bufUsed;
}

TelnetSpy::operator bool() const {
	if (usedSer) {
		return (bool) *usedSer;
	}
	return true;
}

void TelnetSpy::setDebugOutput(bool en) {
	debugOutput = en;
	if (debugOutput) {
		actualObject = this;
#ifdef ESP8266		
		//os_install_putc1((void*) TelnetSpy_putc);  // Set system printing (os_printf) to TelnetSpy
		os_install_putc1( TelnetSpy_putc);  // Set system printing (os_printf) to TelnetSpy // fix base on: https://github.com/yasheena/telnetspy/issues/13
		system_set_os_print(true);
#else // ESP32
		// ToDo: How can be done this for ESP32 ?
#endif
	} else {
		if (actualObject == this) {
#ifdef ESP8266		
			system_set_os_print(false);
			//os_install_putc1((void*) TelnetSpy_ignore_putc); // Ignore system printing
			os_install_putc1( TelnetSpy_ignore_putc); // Ignore system printing  // fix base on: https://github.com/yasheena/telnetspy/issues/13
#else // ESP32
			// ToDo: How can be done this for ESP32 ?
#endif
			actualObject = NULL;
		}
	}
}

uint32_t TelnetSpy::baudRate(void) {
	if (usedSer) {
		return usedSer->baudRate();
	}
	return 115200;
}

void TelnetSpy::sendBlock() {
CRITCAL_SECTION_START
	uint16_t len = bufUsed;
	if (len > maxBlockSize) {
		len = maxBlockSize;
	}
	len = min(len, (uint16_t) (bufLen - bufRdIdx));
	uint16_t idx = bufRdIdx;
CRITCAL_SECTION_END
	client.write(&telnetBuf[idx], len);
CRITCAL_SECTION_START
	bufRdIdx += len;
	if (bufRdIdx >= bufLen) {
		bufRdIdx = 0;
	}
	bufUsed -= len;
	if (bufUsed == 0) {
		bufRdIdx = 0;
		bufWrIdx = 0;
	}
CRITCAL_SECTION_END
	waitRef = 0xFFFFFFFF;
	if (pingRef != 0xFFFFFFFF) {
		pingRef = (millis() & 0x7FFFFFF) + pingTime;
		if (pingRef > 0x7FFFFFFF) {
			pingRef -= 0x80000000;
		}
	}
}

void TelnetSpy::addTelnetBuf(char c) {
CRITCAL_SECTION_START
	telnetBuf[bufWrIdx] = c;
	if (bufUsed == bufLen) {
		bufRdIdx++;
		if (bufRdIdx >= bufLen) {
			bufRdIdx = 0;
		}
	} else {
		bufUsed++;
	}
	bufWrIdx++;
	if (bufWrIdx >= bufLen) {
		bufWrIdx = 0;
	}
CRITCAL_SECTION_END
}

char TelnetSpy::pullTelnetBuf() {
	if (bufUsed == 0) {
		return 0;
	}
CRITCAL_SECTION_START
	char c = telnetBuf[bufRdIdx++]; 
	if (bufRdIdx >= bufLen) {
		bufRdIdx = 0;
	}
	bufUsed--;
CRITCAL_SECTION_END
	return c;
}

char TelnetSpy::peekTelnetBuf() {
	if (bufUsed == 0) {
		return 0;
	}
CRITCAL_SECTION_START
char c = telnetBuf[bufRdIdx]; 
CRITCAL_SECTION_END
//	return telnetBuf[bufRdIdx]; 
return c;
}

int TelnetSpy::telnetAvailable() {
	int n = client.available();
	while (n > 0) {
		if (0xff == client.peek()) {  // If esc char for telnet NVT protocol data remove that telegram:
			client.read();  // Remove esc char
			n--;
			if (0xff == client.peek()) {  // If esc sequence for 0xFF data byte...
				return n; // ...return info about available data (just this 0xFF data byte)
			}
			client.read();  // Skip the rest of the telegram of the telnet NVT protocol data
			client.read();
			n--;
			n--;
		} else {  // If next char is a normal data byte...
			return n;   // ...return info about available data
		}
	}
	return 0;
}

bool TelnetSpy::isClientConnected() {
	return connected;
}

void TelnetSpy::setCallbackOnConnect(void (*callback)()) {
	callbackConnect = callback;
}

void TelnetSpy::setCallbackOnDisconnect(void (*callback)()) {
	callbackDisconnect = callback;
}

void TelnetSpy::handle() {
	if (firstMainLoop) {
		firstMainLoop = false;
    	// Between setup() and loop() the configuration for os_print may be changed so it must be renewed 
		if (debugOutput && (actualObject == this)) {
			setDebugOutput(true);
		}
	}
	if (!started) {
		return;
	}
	if (!listening) {
		if (WiFi.status() != WL_CONNECTED) {
			return;
		}
		telnetServer = new WiFiServer(port);
		telnetServer->begin();
		telnetServer->setNoDelay(bufLen > 0);
		listening = true;
	}
    if (telnetServer->hasClient()) {
        if (client.connected()) {
            WiFiClient rejectClient = telnetServer->available();
			if (strlen(rejectMsg) > 0) {
				rejectClient.write((const uint8_t*) rejectMsg, strlen(rejectMsg));
			}
			rejectClient.flush();
            rejectClient.stop();
        } else {
            client = telnetServer->available();
			if (strlen(welcomeMsg) > 0) {
				client.write((const uint8_t*) welcomeMsg, strlen(welcomeMsg));
			}
        }
    }
    if (client.connected()) {
    	if (!connected) {
    		connected = true;
    		if (pingTime != 0) {
    			pingRef = (millis() & 0x7FFFFFF) + pingTime;
    		}
			if (callbackConnect != NULL) {
				callbackConnect();
			}
		}
	} else {
    	if (connected) {
    		connected = false;
        	client.flush();
            client.stop();
			pingRef = 0xFFFFFFFF;
			waitRef = 0xFFFFFFFF; 
			if (callbackDisconnect != NULL) {
				callbackDisconnect();
			}
		}
	}
	
	if (client.connected() && (bufUsed > 0)) {
		if (bufUsed >= minBlockSize) {
			sendBlock();
		} else {
			unsigned long m = millis() & 0x7FFFFFF;
			if (waitRef == 0xFFFFFFFF) {
				waitRef = m + collectingTime;
				if (waitRef > 0x7FFFFFFF) {
					waitRef -= 0x80000000;
				}
			} else {
				if (!((waitRef < 0x20000000) && (m > 0x60000000)) && (m >= waitRef)) {
					sendBlock();
				}
			}
		}
	}
	if (client.connected() && (pingRef != 0xFFFFFFFF)) {
		unsigned long m = millis() & 0x7FFFFFF;
		if (!((pingRef < 0x20000000) && (m > 0x60000000)) && (m >= pingRef)) {
			addTelnetBuf(0);
			sendBlock();
		}
	}
}


library/SerialCom/SerialCom_non_blocking.h
/*
 SerialCom protocol library - non-blocking.

 Devised and written by Nick Gammon - modified by Radek Vesely for Rep_Al Mower project
 Original library here: https://www.gammon.com.au/forum/?id=11428
 Date: 5 May 2022
 Version: 1.1 - for SW and HW serial

 Licence: Released for public use.

*/


#ifndef SerialCom_non_blocking_h
#define SerialCom_non_blocking_h

#include "Arduino.h"


class SerialCom
  {
  enum {
        STX = '\2',   // start of text
        ETX = '\3'    // end of text
  };  // end of enum

  // where we save incoming stuff
  byte * data_;

  // how much data is in the buffer
  const int bufferSize_;

  // this is true once we have valid data in buf
  bool available_;

  // an STX (start of text) signals a packet start
  bool haveSTX_;

  // count of errors
  unsigned long errorCount_;

  // variables below are set when we get an STX
  bool haveETX_;
  byte inputPos_;
  byte currentByte_;
  bool firstNibble_;
  unsigned long startTime_;

  // helper private functions
  byte crc8 (const byte *addr, byte len);
  void sendComplemented (const byte what);

  Stream *_theSerial; // Somewhere to store the object's pointer

  public:

    // constructor

    // How to write a library that support both HW and SW Serial communication
    // https://www.youtube.com/watch?v=-smhZUVd_0o
    SerialCom(Stream *serial, const byte bufferSize) : _theSerial(serial),  bufferSize_(bufferSize) {}
    SerialCom(Stream &serial, const byte bufferSize) : _theSerial(&serial), bufferSize_(bufferSize) {}

    // destructor - frees memory used
    ~SerialCom () { stop (); }

    // allocate memory for buf_
    void begin ();

    // free memory in buf_
    void stop ();

    // handle incoming data, return true if packet ready
    bool update ();

    // reset to no incoming data (eg. after a timeout)
    void reset ();

    // send data
    void sendMsg (const byte * data, const byte length);

    // returns true if packet available
    bool available () const { return available_; };

    // once available, returns the address of the current message
    byte * getData ()   const { return data_; }
    byte   getLength () const { return inputPos_; }

    // return how many errors we have had
    unsigned long getErrorCount () const { return errorCount_; }

    // return when last packet started
    unsigned long getPacketStartTime () const { return startTime_; }

    // return true if a packet has started to be received
    bool isPacketStarted () const { return haveSTX_; }

  }; // end of class SerialCom

#endif /* SerialCom_non_blocking_h */


library/DFRobot_QMC5883-master/DFRobot_QMC5883.h
/*!
 * @file DFRobot_QMC5883.h
 * @brief Compatible with QMC5883 and QMC5883
 * @n 3-Axis Digital Compass IC
 *
 * @copyright	[DFRobot](http://www.dfrobot.com), 2017
 * @copyright	GNU Lesser General Public License
 *
 * @author [dexian.huang](952838602@qq.com)
 * @version  V1.0
 * @date  2017-7-3
 */

#ifndef DFROBOT_QMC5883_H
#define DFROBOT_QMC5883_H

#if ARDUINO >= 100
#include "Arduino.h"
#else
#include "WProgram.h"
#endif

#define HMC5883L_ADDRESS             (0x1E)
#define QMC5883_ADDRESS              (0x0D)



#define HMC5883L_REG_CONFIG_A         (0x00)
#define HMC5883L_REG_CONFIG_B         (0x01)
#define HMC5883L_REG_MODE             (0x02)
#define HMC5883L_REG_OUT_X_M          (0x03)
#define HMC5883L_REG_OUT_X_L          (0x04)
#define HMC5883L_REG_OUT_Z_M          (0x05)
#define HMC5883L_REG_OUT_Z_L          (0x06)
#define HMC5883L_REG_OUT_Y_M          (0x07)
#define HMC5883L_REG_OUT_Y_L          (0x08)
#define HMC5883L_REG_STATUS           (0x09)
#define HMC5883L_REG_IDENT_A          (0x0A)
#define HMC5883L_REG_IDENT_B          (0x0B)
#define HMC5883L_REG_IDENT_C          (0x0C)


#define QMC5883_REG_OUT_X_M          (0x01)
#define QMC5883_REG_OUT_X_L          (0x00)
#define QMC5883_REG_OUT_Z_M          (0x05)
#define QMC5883_REG_OUT_Z_L          (0x04)
#define QMC5883_REG_OUT_Y_M          (0x03)
#define QMC5883_REG_OUT_Y_L          (0x02)
#define QMC5883_REG_STATUS           (0x06)
#define QMC5883_REG_CONFIG_1         (0x09)
#define QMC5883_REG_CONFIG_2         (0x0A)
#define QMC5883_REG_IDENT_B          (0x0B)
#define QMC5883_REG_IDENT_C          (0x20)
#define QMC5883_REG_IDENT_D          (0x21)


typedef enum
{
  HMC5883L_SAMPLES_8    = 0b11,
  HMC5883L_SAMPLES_4    = 0b10,
  HMC5883L_SAMPLES_2    = 0b01,
  HMC5883L_SAMPLES_1    = 0b00,

  QMC5883_SAMPLES_8     = 0b11,
  QMC5883_SAMPLES_4     = 0b10,
  QMC5883_SAMPLES_2     = 0b01,
  QMC5883_SAMPLES_1     = 0b00
} QMC5883_samples_t;

typedef enum
{
  HMC5883L_DATARATE_75HZ       = 0b110,
  HMC5883L_DATARATE_30HZ       = 0b101,
  HMC5883L_DATARATE_15HZ       = 0b100,
  HMC5883L_DATARATE_7_5HZ      = 0b011,
  HMC5883L_DATARATE_3HZ        = 0b010,
  HMC5883L_DATARATE_1_5HZ      = 0b001,
  HMC5883L_DATARATE_0_75_HZ    = 0b000,

  QMC5883_DATARATE_10HZ        = 0b00,
  QMC5883_DATARATE_50HZ        = 0b01,
  QMC5883_DATARATE_100HZ       = 0b10,
  QMC5883_DATARATE_200HZ       = 0b11
} QMC5883_dataRate_t;

typedef enum
{
  HMC5883L_RANGE_8_1GA     = 0b111,
  HMC5883L_RANGE_5_6GA     = 0b110,
  HMC5883L_RANGE_4_7GA     = 0b101,
  HMC5883L_RANGE_4GA       = 0b100,
  HMC5883L_RANGE_2_5GA     = 0b011,
  HMC5883L_RANGE_1_9GA     = 0b010,
  HMC5883L_RANGE_1_3GA    = 0b001,
  HMC5883L_RANGE_0_88GA    = 0b000,

  QMC5883_RANGE_2GA     = 0b00,
  QMC5883_RANGE_8GA     = 0b01
} QMC5883_range_t;

typedef enum
{
  HMC5883L_IDLE         = 0b10,
  HMC5883_SINGLE        = 0b01,
  HMC5883L_CONTINOUS    = 0b00,

  QMC5883_SINGLE        = 0b00,
  QMC5883_CONTINOUS     = 0b01
} QMC5883_mode_t;

#define IC_NONE     0
#define IC_HMC5883L 1
#define IC_QMC5883  2



#ifndef VECTOR_STRUCT_H
#define VECTOR_STRUCT_H
struct Vector
{
  float XAxis;
  float YAxis;
  float ZAxis;
};
#endif

class DFRobot_QMC5883
{
public:
  DFRobot_QMC5883():isHMC_(false),isQMC_(false),minX(0),maxX(0),minY(0), maxY(0), minZ(0), maxZ(0),firstRun(true)
    {}
  bool begin(void);

  Vector readRaw(void);
  Vector readNormalize(void);
  
  void initMinMax();
  void calibrate(void);
  
  void  setRange(QMC5883_range_t range);
  QMC5883_range_t getRange(void);

  void  setMeasurementMode(QMC5883_mode_t mode);
  QMC5883_mode_t getMeasurementMode(void);

  void  setDataRate(QMC5883_dataRate_t dataRate);
  QMC5883_dataRate_t getDataRate(void);

  void  setSamples(QMC5883_samples_t samples);
  QMC5883_samples_t getSamples(void);
  int getICType(void);
  bool isHMC(){return isHMC_;}
  bool isQMC(){return isQMC_;}
private:
  bool isHMC_;
  bool isQMC_;

  float mgPerDigit;
  Vector v;
  
  
  float minX, maxX;
  float minY, maxY;
  float minZ, maxZ;
  bool firstRun;
  void writeRegister8(uint8_t reg, uint8_t value);
  uint8_t readRegister8(uint8_t reg);
  uint8_t fastRegister8(uint8_t reg);
  int16_t readRegister16(uint8_t reg);
};

#endif


library/SerialCom/SerialCom_non_blocking.cpp
/*
 SerialCom protocol library - non-blocking.

 Devised and written by Nick Gammon - modified by Radek Vesely for Rep_Al Mower project
 Original library here: https://www.gammon.com.au/forum/?id=11428
 Date: 5 May 2022
 Version: 1.1 - for SW and HW serial

 Can send from 1 to 255 bytes from one node to another with:

 * Packet start indicator (STX)
 * Each data byte is doubled and inverted to check validity
 * Packet end indicator (ETX)
 * Packet CRC (checksum)


 To allow flexibility with hardware (eg. Serial, SoftwareSerial, I2C)
 you provide three "callback" functions which send or receive data. Examples are:


 PERMISSION TO DISTRIBUTE

 Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 and associated documentation files (the "Software"), to deal in the Software without restriction,
 including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.


 LIMITATION OF LIABILITY

 The software is provided "as is", without warranty of any kind, express or implied,
 including but not limited to the warranties of merchantability, fitness for a particular
 purpose and noninfringement. In no event shall the authors or copyright holders be liable
 for any claim, damages or other liability, whether in an action of contract,
 tort or otherwise, arising from, out of or in connection with the software
 or the use or other dealings in the software.

 */


#include <SerialCom_non_blocking.h>


// allocate the requested buffer size
void SerialCom::begin ()
  {
  data_ = (byte *) malloc (bufferSize_);
  reset ();
  errorCount_ = 0;
  } // end of SerialCom::begin

// get rid of the buffer
void SerialCom::stop ()
{
  reset ();
  free (data_);
  data_ = NULL;
} // end of SerialCom::stop

// called after an error to return to "not in a packet"
void SerialCom::reset ()
  {
  haveSTX_ = false;
  available_ = false;
  inputPos_ = 0;
  startTime_ = 0;
  } // end of SerialCom::reset

// calculate 8-bit CRC
byte SerialCom::crc8 (const byte *addr, byte len)
{
  byte crc = 0;
  while (len--)
    {
    byte inbyte = *addr++;
    for (byte i = 8; i; i--)
      {
      byte mix = (crc ^ inbyte) & 0x01;
      crc >>= 1;
      if (mix)
        crc ^= 0x8C;
      inbyte >>= 1;
      }  // end of for
    }  // end of while
  return crc;
}  // end of SerialCom::crc8

// send a byte complemented, repeated
// only values sent would be (in hex):
//   0F, 1E, 2D, 3C, 4B, 5A, 69, 78, 87, 96, A5, B4, C3, D2, E1, F0
void SerialCom::sendComplemented (const byte what)
{
byte c;

  // first nibble
  c = what >> 4;
  _theSerial->write ((c << 4) | (c ^ 0x0F));

  // second nibble
  c = what & 0x0F;
  _theSerial->write ((c << 4) | (c ^ 0x0F));

}  // end of SerialCom::sendComplemented

// send a message of "length" bytes (max 255) to other end
// put STX at start, ETX at end, and add CRC
void SerialCom::sendMsg (const byte * data, const byte length)
{
  _theSerial->write (STX);  // STX
  for (byte i = 0; i < length; i++)
    sendComplemented (data [i]);
  _theSerial->write (ETX);  // ETX
  sendComplemented (crc8 (data, length));
}  // end of SerialCom::sendMsg

// called periodically from main loop to process data and
// assemble the finished packet in 'data_'

// returns true if packet received.

// You could implement a timeout by seeing if isPacketStarted() returns
// true, and if too much time has passed since getPacketStartTime() time.

bool SerialCom::update ()
  {
  // no data? can't go ahead (eg. begin() not called)
  if (data_ == NULL)
     return false;

  while (_theSerial->available () > 0)
    {
    byte inByte = _theSerial->read ();
    switch (inByte)
      {

        case STX:   // start of text
          haveSTX_ = true;
          haveETX_ = false;
          inputPos_ = 0;
          firstNibble_ = true;
          startTime_ = millis ();
          break;

        case ETX:   // end of text (now expect the CRC check)
          haveETX_ = true;
          break;

        default:
          // wait until packet officially starts
          if (!haveSTX_)
            break;

          // check byte is in valid form (4 bits followed by 4 bits complemented)
          if ((inByte >> 4) != ((inByte & 0x0F) ^ 0x0F) )
            {
            reset ();
            errorCount_++;
            break;  // bad character
            } // end if bad byte

          // convert back
          inByte >>= 4;

          // high-order nibble?
          if (firstNibble_)
            {
            currentByte_ = inByte;
            firstNibble_ = false;
            break;
            }  // end of first nibble

          // low-order nibble
          currentByte_ <<= 4;
          currentByte_ |= inByte;
          firstNibble_ = true;

          // if we have the ETX this must be the CRC
          if (haveETX_)
            {
            if (crc8 (data_, inputPos_) != currentByte_)
              {
              reset ();
              errorCount_++;
              break;  // bad crc
              } // end of bad CRC

            available_ = true;
            return true;  // show data ready
            }  // end if have ETX already

          // keep adding if not full
          if (inputPos_ < bufferSize_)
        	  data_ [inputPos_++] = currentByte_;
          else
            {
            reset (); // overflow, start again
            errorCount_++;
            }

          break;

      }  // end of switch
    }  // end of while incoming data

  return false;  // not ready yet
  } // end of SerialCom::update

